### ISSUE TITLE
cli: `cockroach node status` is broken

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#43974]
Found while working on #43893  

- `cockroach node status` is meant to work even if the node is unhealthy (eg. a system range is unavailable), as per a comment in roachtests and my own (knz') personal recollections of past discussions
- _however_ at the same time its is implemented using a SQL query
- SQL is not meant to work when system.users is unavailable, because every connection should authenticate the user logging in

Therefore _currently_ `cockroach node status` happens to work for the `root` user because of some bypass in the code when the username is "`root`" exactly. However, it is broken if used with any other client cert than `root`'s.

There are two perspectives about this:

A. either we consider that SQL should always authenticate (which creates consistency and regularity), in which case `cockroach node status` should be made to use the status RPC entirely and bypass SQL.

   For this to be viable we must start to commit to (and better understand) a separation between "control" (RPC) and "data" (SQL) planes in the APIs.

B. or, we should consider that the "main" administration API for crdb is SQL and that this admin API should always work even when system ranges are unavailable, i.e. we must implement and document a way for clients to establish SQL connections when the system ranges are unavailable.

   For this to be viable we need a special "operator" account that's always special in this way, but without removing the ability for users to use a non-special account that's also admin in the UI (i.e. properly address #43847 / #43870) 

(needless say my(knz) heart goes towards direction A)

**Comment by knz** [comment]
@tbg can you check the above? I'd like to point out that `node` was using the RPC originally and that you specifically required SQL in #20713. 

Interestingly, #20713 was listing `cockroach user`, `zone` and `node`. Both `user` and `zone` have been removed since because they were redundant with SQL.

It seems to me that `node` fundamentally should not be done via SQL, and that essentially the thrust behind #20713 should have been narrowed to `user` and `zone`, and thus the current behavior reverted for `node`. Do you agree?

**Comment by ajwerner** [comment]
I can see some value in `B.` though I haven't decided if I'd go so far as to say I like it. In particular, shoving more administrative functionality underneath SQL gives us a smaller API surface w.r.t. API uniformity and concepts for users. I'm generally in favor of exposing non-performance critical functionality to users via a SQL (crdb_internal) interface rather than a new gRPC service. I anticipate that it can ease client adoption if that client has gone through the work of setting up a connection. I also suspect that there are use cases where it is convenient to be able to access data available from other APIs via SQL.

As noted, we'd need to tell a new story about authn/authz for such a sql user. Perhaps we want to provide a mechanism to create tokens with an expiration and well scoped access to "control" data. Furthermore, pushing such data underneath SQL will muddle the interface.

Perhaps my preference is to do both. Define a clear "control" portion which is implemented on top of SQL and can bypass authz explicitly. Then, with less urgency, provide a mechanism to authorize SQL connections to access the same data.

**Comment by knz** [comment]
> Perhaps we want to provide a mechanism to create tokens with an expiration and well scoped access to "control" data

Yeah and that info will be stored ... where? In a KV range that needs quorum to access? That sounds like a snake biting its tail.

I'm cooking a high-level description of the problem that underlines how we've been designing with conflicting goals. There will be fallout.

**Comment by ajwerner** [comment]
> Yeah and that info will be stored ... where? In a KV range that needs quorum to access? That sounds like a snake biting its tail.

In my head the proposal was that the token be irrevocable or best effort revocable. The idea was that it could be a signed token which stores its expiration so that verification can occur without access to the database.

**Comment by knz** [comment]
> idea was that it could be a signed token which stores its expiration so that verification can occur without access to the database.

We have that already with TLS client certs (signed + embedded expiration)

Except that it's not server-side revocable. And we can't control/revoke privileges.

**Comment by tbg** [comment]
None of the options discussed seems very appealing. Take a look at the SQL used for `node status`:

https://github.com/cockroachdb/cockroach/blob/0917dcc9cf61f060d67c328131a7a73e136109a8/pkg/cli/node.go#L148-L189

Cobbling this together without SQL is unsavory. In fact, the previous code doing so was... bad, we can't revert to it.

The two options I find palatable are:

1. don't do anything (which I suppose means that root can't have a password) - I bring this up because I wasn't able to find a clear reason why we wanted to make that change in the first place. We may thus have the option to fight this fight another day.
2. push the code for the administrative commands out of the cli into the server. That is, https://github.com/cockroachdb/cockroach/blob/0917dcc9cf61f060d67c328131a7a73e136109a8/pkg/cli/node.go#L129 would do nothing but open a gRPC connection and send an RPC containing the args, whose output is then rendered without any additional logic. The server side could create an "internal" authenticated session that can run SQL, something I believe we already do in some places.

**Comment by knz** [comment]
I like 2 as it's well aligned with another item on the CLI roadmap. I'll mull it over.

**Comment by knz** [comment]
RFC for a solution: https://github.com/knz/cockroach/blob/20200119-users-rfc/docs/RFCS/20200119_always_available_users.md
